### PR TITLE
Add method to detect disputable liquidations

### PR DESCRIPTION
**Pull request by ptare**
The API for this is a bit clunky, I don't love it.

Issue #939 

Signed-off-by: Prasad Tare <prasad.tare@gmail.com>

### CODE REVIEW

**Comment by ptare** [comment]
> `getUnderCollateralizedPositions` looks good, a few higher level questions:
> 
> * How do the class variables `sponsorAddresses` and `positions` get filled? Or will they be passed in by the caller of the client

In `_update()`.

> * What don't you love about the API? Happy to brainstorm some ideas if you have some general inefficiencies or complaints about the structure in mind

There's no good way to get all the disputable liquidations. With the current API, you'd first get all the undisputed liquidations, then check all of them if they are disputable.

That happens because you need to first get the liquidation times in order to get the price at those times.

**Comment by nicholaspai** [comment]
> > `getUnderCollateralizedPositions` looks good, a few higher level questions:
> > 
> > * How do the class variables `sponsorAddresses` and `positions` get filled? Or will they be passed in by the caller of the client
> 
> In `_update()`.
Ah missed that.
> 
> > * What don't you love about the API? Happy to brainstorm some ideas if you have some general inefficiencies or complaints about the structure in mind
> 
> There's no good way to get all the disputable liquidations. With the current API, you'd first get all the undisputed liquidations, then check all of them if they are disputable.
> 
> That happens because you need to first get the liquidation times in order to get the price at those times.

So the only way I can think of to get all liquidation times would be to read the event logs, does that work?

File: financial-templates-lib/ExpiringMultiPartyClient.js [hunk 1]
```diff
@@ -21,27 +21,22 @@ class ExpiringMultiPartyClient {
...
   // Returns an array of { sponsor, id, numTokens, amountCollateral, liquidationTime } for each undisputed liquidation.
+  // To check whether a liquidation can be disputed, call `isDisputable` with the token redemption value at
+  // `liquidationTime`.
   getUndisputedLiquidations = () => this.undisputedLiquidations;
 
+  // Whether the given `liquidation` (`getUndisputedLiquidations` returns an array of `liquidation`s) is disputable.
+  // `tokenRedemptionValue` should be the redemption value at `liquidation.time`.
+  isDisputable = (liquidation, tokenRedemptionValue) => {
+    return !this._isUnderCollateralized(liquidation.numTokens, liquidation.amountCollateral, tokenRedemptionValue);
+  };
```
Comment (mrice32) [inline financial-templates-lib/ExpiringMultiPartyClient.js]: Should we check here or in the intake step of `_update()` that the liquidation isn't past its expiry/liveness period?

File: financial-templates-lib/test/ExpiringMultiPartyClient.js [hunk 2]
```diff
@@ -114,5 +127,25 @@ contract("ExpiringMultiPartyClient.js", function(accounts) {
...
+
+    // Create a new liquidation for account[0]'s position.
+    const id = await emp.createLiquidation.call(accounts[0], { rawValue: toWei("150") }, { from: accounts[1] });
+    await emp.createLiquidation(accounts[0], { rawValue: toWei("150") }, { from: accounts[1] });
+    await client._update();
+
+    const liquidations = client.getUndisputedLiquidations();
+    // Disputable if the disputer believes the price was `1`, and not disputable if they believe the price was just
+    // above `1`.
+    assert(client.isDisputable(liquidations[0], toWei("1")));
```
Comment (mrice32) [inline financial-templates-lib/test/ExpiringMultiPartyClient.js]: nit: why not `assert.isTrue`?

File: financial-templates-lib/ExpiringMultiPartyClient.js (hunk 1 above)
Comment (nicholaspai) [reply financial-templates-lib/ExpiringMultiPartyClient.js]: + or either has already been disputed or is pending dispute

Comment (mrice32) [reply financial-templates-lib/ExpiringMultiPartyClient.js]: > ++ or either has already been disputed or is pending dispute

I think that's already checked on line 93 where the liquidations are pulled in.

File: financial-templates-lib/test/ExpiringMultiPartyClient.js (hunk 2 above)
Comment (ptare) [reply financial-templates-lib/test/ExpiringMultiPartyClient.js]: Done.

File: financial-templates-lib/ExpiringMultiPartyClient.js (hunk 1 above)
Comment (ptare) [reply financial-templates-lib/ExpiringMultiPartyClient.js]: Yep, we should definitely check for that. Done in _update, thoughts?

**Comment by ptare** [comment]
> > > `getUnderCollateralizedPositions` looks good, a few higher level questions:
> > > 
> > > * How do the class variables `sponsorAddresses` and `positions` get filled? Or will they be passed in by the caller of the client
> > 
> > 
> > In `_update()`.
> > Ah missed that.
> > > * What don't you love about the API? Happy to brainstorm some ideas if you have some general inefficiencies or complaints about the structure in mind
> > 
> > 
> > There's no good way to get all the disputable liquidations. With the current API, you'd first get all the undisputed liquidations, then check all of them if they are disputable.
> > That happens because you need to first get the liquidation times in order to get the price at those times.
> 
> So the only way I can think of to get all liquidation times would be to read the event logs, does that work?

The difficult part is accessing the price feed, so we'd still need something weird here. Happy to revisit this as necessary.

**Pull request merged**